Add initial AI crawler post (#317)
readthedocs/website: content/posts/ai-crawlers-abuse.md

title: AI Crawlers need to be more respectful
date: 2024-07-25
category: Engineering
tags: ai, crawlers, abuse
authors: Eric Holscher
status: published
image:
image_credit:

> Many AI companies are starting to crawl many pages on the web,

Reviewer: So, overall, I think this post would benefit from describing the problem we're facing sooner. The end of this post did a great job of that and especially describing our frustration.

Just tossing out an example, but maybe this first sentence just gets right to the point:

```suggestion
In the last few months, we have noticed an increase in abusive site crawling,
mainly from AI products and services. These products are recklessly crawling
many sites across the web, and we've already had to block several sources of abusive traffic.
```

trying to extract as much information as fast as possible.
It feels like a new gold rush,
and in their haste,

> some of these crawlers are being so badly they are causing harm to the sites they are crawling.

Reviewer: ```suggestion
some of these crawlers are behaving so badly that they are causing harm to the sites they are crawling.
```

At Read the Docs,
we host documentation for many projects and are generally bot friendly,
but the behavior of AI crawlers is currently causing us problems.

> We host HTML zip files (``htmlzip``) of documentation for many projects,
> and these can be somewhat large.
> The AI Crawlers are downloading these files repeatedly,
> and without any rate or bandwidth limiting.

Reviewer: What is actually happening is explained well below. Introducing the problem at a high level would flow nicely here. Again, just an idea of what might work:

```suggestion
We have noticed AI crawlers aggressively pulling content, seemingly without basic
checks against abuse. Bots repeatedly download large files hundreds of
times in a single day, with traffic sent from distributed sources without
rate limiting or bandwidth limitations.
```

AI crawlers are acting in a way that is not respectful to the sites they are crawling,
and that is going to cause a backlash against AI crawlers in general.
As a community-supported site without a large budget,
AI crawlers have cost us a significant amount of money in bandwidth charges,
and caused us to spend a large amount of time dealing with abuse.

## AI crawler abuse

> I want to start off with a few examples of the types of abuse we are seeing:

Reviewer: Only because we're half way through the post:

```suggestion
Here are a few examples of the types of abuse we are seeing:
```

### 73 TB in March 2024 from one crawler

> **One crawler downloaded 73 TB of htmlzip files in May 2024, with almost 10 TB in a single day**. This cost us over $5,000 in bandwidth charges, and we had to block the crawler. We emailed this company, reporting a bug in their crawler, and we're working on them reimbursing us for the costs.

Reviewer: I mentioned above we explained the problem well down here, but it might be worth moving the description of these files here. I would definitely mention these files, you mention below they don't change often too, which maybe helps with context here.

Reviewer: ```suggestion
**One crawler downloaded 73 TB of htmlzip files in May 2024, with almost 10 TB in a single day**. This cost us over $5,000 in bandwidth charges, and we had to block the crawler. We emailed this company, reporting a bug in their crawler, and we're working with them on reimbursing us for the costs.
```

![AI Crawler abuse, May 2024](/images/posts/bandwidth-may-2024.png)

This was a bug in their crawler that was causing it to download the same files over and over again.
There was no bandwidth limiting in place,
or support for Etags and Last-Modified headers which would have allowed the crawler to only download files that had changed.
We have reported this issue to them,
and hopefully the issue will be fixed.

We do have a CDN for these files,
but this request was for an old URL that had an old redirect in place.
This redirect went to an old dashboard download URL,
where we don't have CDN caching in place for security reasons around serving other dynamic content.
We are planning to fix this redirect to point to the cached URL.

<details>
<summary>Example web requests</summary>

<code>

    -> curl -IL "https://media.readthedocs.org/htmlzip/chainer/v1.24.0/_modules/chainer/testing/_modules/chainer/_modules/cupy/indexing/_modules/chainer/initializers/normal.html"
    HTTP/2 302
    date: Thu, 25 Jul 2024 16:24:18 GMT
    content-type: text/html
    content-length: 138

>     location: https://buildmedia.readthedocs.org/media/htmlzip/chainer/v1.24.0/_modules/chainer/testing/_modules/chainer/_modules/cupy/indexing/_modules/chainer/initializers/normal.html

Reviewer 2: BTW, this doesn't render nicely unfortunately: https://readthedocs-about--317.org.readthedocs.build/blog/2024/07/ai-crawlers-abuse/

Reviewer 2: I think you can use just triple double ticks here.

    content-length: 5888860
    content-disposition: filename=docs-chainer-org-en-v1.24.0.zip
    x-amz-id-2: +DjI2tMbUou9XNK5+G53Gyhah4lhBwAgnRiqBh9vsR3KzqxajSTC4B+eIQBY+pi+ZR6McRQngSI=
    x-amz-request-id: Z502YT87WEMM3ZY9
    last-modified: Thu, 11 Feb 2021 09:12:59 GMT
    etag: "c8cb418f5a8ff2e376fc5f7b7564e445"
    x-amz-meta-mtime: 1495712654.422637991
    accept-ranges: bytes
    x-served: Nginx-Proxito-Sendfile
    x-backend: web-i-01ce033e08bb601ef
    referrer-policy: strict-origin-when-cross-origin
    x-frame-options: DENY
    x-content-type-options: nosniff
    content-security-policy: object-src 'none'; frame-ancestors 'none'
    cf-cache-status: DYNAMIC
    set-cookie: __cf_bm=qGTC_35C03_QI6rw.JyPZhFHpo2QxUy7DMMFJpjz2_U-1721924658-1.0.1.1-4iS9rZHPJmt_I5rQX4NuKr_pHQmCw0jvCzAIYX.CeUtGZh6hIZIjBWhlPoxEMjhsRcvbuTSpgSa9oltlvYDtEA; path=/; expires=Thu, 25-Jul-24 16:54:18 GMT; domain=.readthedocs.org; HttpOnly; Secure; SameSite=None
    set-cookie: _cfuvid=buQRaZWXJEn51CIpRsDW3E52DyDqEHd_sgY5PxOLfyE-1721924658779-0.0.1.1-604800000; path=/; domain=.readthedocs.org; HttpOnly; Secure; SameSite=None
    server: cloudflare
    cf-ray: 8a8d7fdbbf787565-SEA

</code>

> As you can see, this file was last modified in 2021,
> but it was downloaded hundreds of times.

Reviewer: Just to separate these two statements so it doesn't imply "As you can see it was downloaded hundreds of times":

```suggestion
As you can see, this file was last modified in 2021.
However, crawlers didn't have this basic check and instead repeatedly downloaded files like these hundreds of times each.
```

</details>

### 10 TB in June 2024 from another

In June 2024, **someone used Facebook's content downloader to download 10 TB** of data, mostly htmlzip and PDF files. When we tried to email Facebook about it with the [contact information listed in the bot's user agent](http://www.facebook.com/externalhit_uatext.php), but the email bounced.

![AI Crawler abuse, June 2024](/images/posts/bandwidth-june-2024.png)

We think this was someone using Facebook's content downloader to download all of our files, but we aren't really sure.

> It would also be Meta crawling the site for their latest AI project,

Reviewer: ```suggestion
It could also be Meta crawling sites with their latest AI project,
```

but the user agent is not clear.

## Actions taken

We have IP-based rate limiting in place for many of our endpoints,
however these crawlers are coming from a large number of IP addresses,
so our rate limiting is not effective.

These crawlers are using real user agents,
which is a good thing and allows us to identify them,
but it's very difficult for us to block on user agent because many real users use the same browsers with the same user agent.

We have taken a few actions to try to mitigate this abuse:

* We have temporarily blocked all traffic from bots [Cloudflare identifies as AI Crawlers](https://radar.cloudflare.com/traffic/verified-bots), while we figure out how to deal with this.
* We have reconfigured our CDN to better cache these files, so that our origin servers are not hit as hard.
* We have started monitoring our bandwidth usage more closely and are working on more aggressive rate limiting rules.

## Outcomes

Given that our Community site is only for hosting open source projects,
AWS and Cloudflare do give us sponsored plans,
but we only have a limited number of credits each year.

> These bandwidth costs mean that we will run out of AWS credits.

Reviewer: Because it's been a few paragraphs since we discussed bandwidth costs:

```suggestion
The additional bandwidth costs AI crawlers are currently causing will likely mean we will run out of AWS credits early.
```

bandwidth for our downloaded files has decreased by 75% (~800GB/day to ~200GB/day).
If all this traffic hit our origins servers,
it would cost around $50/day, or $1,500/month,
along with the increased load on our servers.

Normal traffic gets cached by our CDN,
and doesn't cost us anything for bandwidth.
But because many of these files are not downloaded often,
as they are scraped the cache is expired and the requests hit our origin servers directly,
causing substantial bandwidth charges.

## Next steps

We are asking all AI companies to be more respectful of the sites they are crawling.
They are risking many sites blocking them for abuse,
irrespective of the other copyright and moral issues that are at play in the industry.

As a large host of open source documentation,
we'd also love to work with these companies on deal to be able to crawl our site in a respectful way.
We could build an integration that would alert them to content changes,
and download the files that have changed.
However, none of these companies have reached out to us in any way,
expect in response to abuse reports.

If these companies wish to be good actors in the space,
they need to start acting like it,

> instead of burning bridges with folks in the community.

Reviewer: This last section is great :100:
